tls/s2n_connection_serialize.c

```diff
@@ -0,0 +1,284 @@
+        POSIX_GUARD(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+        *length = S2N_SERIALIZED_CONN_FIXED_SIZE + (secret_size * 3);
+    } else {
+        *length = S2N_SERIALIZED_CONN_TLS12_SIZE;
+    }
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_connection_serialize_tls13_secrets(struct s2n_connection *conn, struct s2n_stuffer *output)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+
+    uint8_t secret_size = 0;
+    RESULT_GUARD_POSIX(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->secrets.version.tls13.client_app_secret,
+            secret_size));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->secrets.version.tls13.server_app_secret,
+            secret_size));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->secrets.version.tls13.resumption_master_secret,
+            secret_size));
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_connection_serialize_secrets(struct s2n_connection *conn, struct s2n_stuffer *output)
+{
+    RESULT_ENSURE_REF(conn);
+
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->secrets.version.tls12.master_secret,
+            S2N_TLS_SECRET_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->handshake_params.client_random,
+            S2N_TLS_RANDOM_DATA_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(output, conn->handshake_params.server_random,
+            S2N_TLS_RANDOM_DATA_LEN));
+    return S2N_RESULT_OK;
+}
+
+int s2n_connection_serialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+    POSIX_ENSURE_REF(conn->secure->cipher_suite);
+    POSIX_ENSURE_REF(conn->config);
+
+    /* This version check is meant as a safeguard against new TLS features which might be required post-handshake.
+     * For example, assume that a new TLS parameter is released which affects how data is sent
+     * post-handshake. This parameter must be available in both the s2n-tls version that serializes the 
+     * connection, as well as the version that deserializes the connection. If not, the serializer
+     * may negotiate this feature with its peer, which would cause the deserializer to run into errors
+     * sending data to the peer, if it had not yet been updated.
+     * This kind of version-mismatch can happen during deployments and rollbacks, and therefore we require
+     * the user to tell us which serialized version they support pre-handshake. That way, if a new feature
+     * comes out, we will hold off on negotiating it until the user requests the new serialized connection
+     * version the new feature is tied to (i.e. the request indicates they have finished deploying
+     * the new feature to their entire fleet.)
+     */
```

Comment:
Should this comment be on the field in s2n_config instead?

This is also a pretty big chunk of information. Some whitespace might break it up and make people more likely to read it rather than skip it.